test/counter_server_example/test/counter_server_example.erl

```diff
@@ -0,0 +1,361 @@
+
+%% @doc Client making a second request if the first was successful and
+%% verifying that the second answer is greater than the first.
+valid_client_no_vnet() ->
+  Res1 = valid_local_request(),
+  case Res1 =:= error of
+    true -> ok;
+    false ->
+      Res2 = valid_local_request(),
+      check_results(Res1, Res2)
+  end.
+
+valid_local_request() ->
+  local_request(increment).
+
+%%%-------------------------------------------------------------------
+
+local_request(Request) ->
+  try
+    counter_server:request(?SERVER, Request)
+  catch
+    _:_ -> error
+  end.
+
+%%%-------------------------------------------------------------------
+
+check_results(error, _Res2) -> ok;
+check_results(_Res1, error) -> ok;
+check_results({ok, Gen1, Res1}, {ok, Gen2, Res2}) ->
+  ?assert(Res2 > Res1),
+  ?assert(Gen2 >= Gen1);
+check_results(Res1, Res2) ->
+  ?assertEqual(ok, {unexpected, Res1, Res2}).
+
+%%%-------------------------------------------------------------------
+
+%% @doc
+%% Simulating a scenario with two clients, making requests and
+%% checking results, without using any vnet logic.
+%% The second client will break the server.
+%% Nevertheless, the replies that the first client receives are
+%% expected to satisfy the correctness properties.
+invalid_no_vnet_scenario() ->
+  {ok, P1} = counter_server_supervisor:start_link_no_vnet(),
+  unlink(P1),
+  spawn(fun valid_client_no_vnet/0),
+  spawn(fun invalid_client_no_vnet/0).
+
+%%--------------------------------------------------------------------
+
+invalid_client_no_vnet() ->
+  ?assertEqual(error, local_request(invalid)).
+
+%%%-------------------------------------------------------------------
+
+%% @doc
+%% Same as before, but now the first client checks a property that is
+%% not always satisfied: each reply will be from the same generation
+%% of server. As the server might be broken and restarted, this
+%% assertion will fail in some interleavings.
+invalid_same_gen_no_vnet_scenario() ->
+  {ok, P1} = counter_server_supervisor:start_link_no_vnet(),
+  unlink(P1),
+  spawn(fun valid_client_same_gen_no_vnet/0),
+  spawn(fun invalid_client_no_vnet/0).
+
+%%--------------------------------------------------------------------
+
+valid_client_same_gen_no_vnet() ->
+  Res1 = valid_local_request(),
+  case Res1 =:= error of
+    true -> ok;
+    false ->
+      Res2 = valid_local_request(),
+      check_results(Res1, Res2),
+      check_same_gen(Res1, Res2)
+  end.
+
+%%--------------------------------------------------------------------
+
+check_same_gen(error, _) -> ok;
+check_same_gen(_, error) -> ok;
+check_same_gen({ok, G1, _R1}, {ok, G2, _R2}) ->
+  ?assertEqual(G1, G2);
+check_same_gen(Res1, Res2) ->
+  ?assertEqual(ok, {unexpected, Res1, Res2}).
+
+%%%===================================================================
+
+%% @doc
+%% Simulating a scenario with one node as client, making a request.
+once_valid_scenario() ->
+  Nodes = [server, client],
```

Comment:
Maybe the name of the nodes should be defined as macros too for readability. E.g. `{via, vnet, {?server_node, ?server}}`.